python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,346 @@
+    """
+    Interface for a resource scheduler.
+
+    Implements the `instance_manager.proto ResourceSchedulerService` interface.
+    """
+
+    @abstractmethod
+    def schedule(self, request: SchedulingRequest) -> SchedulingReply:
+        """
+        Given the resource requests and the current cluster state, calculate the
+        target cluster shape by trying to schedule the resource requests on the
+        nodes.
+        """
+        pass
+
+
+class SchedulingNodeStatus(Enum):
+    """
+    The status of a scheduling node (`SchedulingNode`)
+    """
+
+    # The node is to be launched.
+    TO_LAUNCH = "TO_LAUNCH"
+    # The node is pending to be launched (a pending request, or a pending instance)
+    PENDING = "PENDING"
+    # The node is running.
+    RUNNING = "RUNNING"
+
+
+@dataclass
+class SchedulingNode:
+    """
+    A abstraction of a node that can be scheduled on by the resource scheduler.
+
+    A scheduling node is expected to be used as:
+
+        node  = SchedulingNode.from_node_config(node_config)
+        score, infeasible = node.try_schedule(requests)
+
+        .... do something with the score ....
+
+    NOTE:
+        One could also extend the scheduling behavior by overriding `try_schedule`
+    """
+
+    # Class level node id counter.
+    _next_id = 0
+
+    @classmethod
+    def next_id(cls) -> int:
+        """
+        Return the next id to be used for creating a new scheduling node id.
+        """
+        cls._next_id += 1
+        return cls._next_id
+
+    @classmethod
+    def from_node_config(
+        cls,
+        node_config: NodeTypeConfig,
+        status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH,
+    ) -> "SchedulingNode":
+        """
+        Create a scheduling node from a node config.
+        """
+        return cls(
+            node_id=cls.next_node_id(),
+            node_type=node_config.name,
+            total_resources=dict(node_config.resources),
+            available_resources=dict(node_config.resources),
+            labels=dict(node_config.labels),
+            status=status,
+        )
+
+    # A unique id simply for differentiating nodes. Not the same as a normal ray node
+    # id.
+    id: int
+    # Node type name.
+    node_type: NodeType
+    # Requests committed to be placed on this node.
+    sched_requests: List[ResourceRequest] = field(default_factory=list)
+    # The node's current resource capacity.
+    total_resources: Dict[str, float] = field(default_factory=dict)
+    # The node's available resources.
+    available_resources: Dict[str, float] = field(default_factory=dict)
+    # Node's labels, including static or dynamic labels.
+    labels: Dict[str, str] = field(default_factory=dict)
+    # Status
+    status: SchedulingNodeStatus = SchedulingNodeStatus.TO_LAUNCH
+    # Observability descriptive message for why the node was launched in the
+    # first place.
+    # TODO
+    launch_reason: Optional[str] = None
+
+    def try_schedule(
```

Comment:
Combine try_schedule and compute_score

Reply:
yeah - it's combined (try_schedule is returning a score)